### PR TITLE
Batch the FlowSOM process

### LINKED ISSUE #384: Batch processing/batch normalization for SOM
**Is your feature request related to a problem? Please describe.**
The current SOM code iterates through each point one at a time. I think there might be some speedups if we can batch this for multiple points. However, more importantly, I think we're less likely to run into problems with the normalization if we have data from multiple FOVs. 99.9% normalization from a single FOV could lead to weird results. 

**Describe the solution you'd like**
An option to specify a bath size, both for training and running, which controls how much data is fed through the SOM at a time.

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Addresses and closes #384. A lot of the processes are currently being done per `fov`. Instead, we transition to a "per set of `fov`" workflow.

**How did you implement your changes**

- Preprocessing: instead of reading in the entire image array, read in just a few `fovs` at a time
- Clustering: run the trained SOM on multiple `fovs` at a time rather than on a single `fov`.

**Remaining issues**

See the issue page.

### CODE REVIEW

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks nice, some minor tweaks and suggestions only

File: ark/phenotyping/som_utils.py [hunk 1]
```diff
@@ -108,65 +197,32 @@ def create_pixel_matrix(img_xr, seg_labels, base_dir,
...
+            img_xr = load_utils.load_imgs_from_mibitiff(
+                tiff_dir, mibitiff_files=fov_batch, channels=channels, dtype="int16"
+            )
+        else:
+            img_xr = load_utils.load_imgs_from_tree(
+                tiff_dir, fovs=fov_batch, channels=channels, dtype="int16"
+            )
+
+        # iterate over each fov in fov_batch and create pixel matrix accordingly
+        for fov in fov_batch:
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply ark/phenotyping/som_utils.py:219]: Guess not, probably best to make the batch size 1 all the time.

**Comment by alex-l-kong** [comment]
I'm holding off on mibitiff testing for now. At some point, we'd probably need to address that, but for now, the only cohorts we're testing on are standard tiffs.

File: ark/phenotyping/som_utils.py [hunk 2]
```diff
@@ -63,17 +64,20 @@ def compute_cluster_avg(fovs, channels, base_dir,
...
-        img_xr (xarray.DataArray):
-            Array representing image data for each fov
+        fov (str):
+            Name of the fov to index
+        channels (list):
+            List of channels to subset over
+        img_data (numpy.ndarray):
+            Array representing image data for one fov
         seg_labels (xarray.DataArray):
             Array representing segmentation labels for each image
```
Comment (ngreenwald) [inline ark/phenotyping/som_utils.py:82]: docstring should be updated

File: ark/phenotyping/som_utils.py [hunk 3]
```diff
@@ -108,65 +191,32 @@ def create_pixel_matrix(img_xr, seg_labels, base_dir,
...
+            img_xr = load_utils.load_imgs_from_mibitiff(
+                tiff_dir, mibitiff_files=[fov], channels=channels, dtype="int16"
+            )
+        else:
+            img_xr = load_utils.load_imgs_from_tree(
+                tiff_dir, fovs=[fov], channels=channels, dtype="int16"
+            )
+
+        # load segmentation labels in for fov
+        seg_labels = load_utils.load_imgs_from_dir(
```
Comment (ngreenwald) [inline ark/phenotyping/som_utils.py:207]: We can probably just load the raw image. I don't think we need it in an xarray for any reason, we just .values it immediately in the subsequent function

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Just a couple small tweaks

File: ark/phenotyping/som_utils.py (hunk 2 above)
Comment (ngreenwald) [reply ark/phenotyping/som_utils.py:82]: This still says for each image, but it's just for a single image

File: ark/phenotyping/som_utils.py [hunk 4]
```diff
@@ -63,27 +64,107 @@ def compute_cluster_avg(fovs, channels, base_dir,
...
+    # convert into a dataframe
+    pixel_mat = pd.DataFrame(pixel_mat, columns=channels)
+
+    # assign metadata about each entry
+    pixel_mat['fov'] = fov
+    pixel_mat['row_index'] = np.repeat(range(img_data.shape[0]), img_data.shape[1])
+    pixel_mat['column_index'] = np.tile(range(img_data.shape[0]), img_data.shape[1])
+
+    # assign segmentation label
+    # seg_labels_flat = seg_labels.loc[fov, ...].values.flatten()
```
Comment (ngreenwald) [inline ark/phenotyping/som_utils.py:114]: Can remove this line

File: ark/phenotyping/som_utils.py [hunk 5]
```diff
@@ -63,27 +64,107 @@ def compute_cluster_avg(fovs, channels, base_dir,
...
+
+    # normalize each row by total marker counts to convert into frequencies
+    pixel_mat.loc[:, channels] = pixel_mat.loc[:, channels].div(
+        pixel_mat.loc[:, channels].sum(axis=1), axis=0)
+
+    # subset the pixel matrix for training
+    pixel_mat_subset = pixel_mat.sample(frac=subset_proportion, random_state=seed)
+
+    # write complete dataset to feather, needed for cluster assignment
+    feather.write_dataframe(pixel_mat,
```
Comment (ngreenwald) [inline ark/phenotyping/som_utils.py:129]: If we have this function return the pixel_mat and pixel_mat_subset, rather than save to disk, we don't have to pass in the paths to the directories. Then the top level function can do the saving

File: ark/phenotyping/som_utils.py [hunk 6]
```diff
@@ -108,65 +192,35 @@ def create_pixel_matrix(img_xr, seg_labels, base_dir,
...
+            )
+        else:
+            img_xr = load_utils.load_imgs_from_tree(
+                tiff_dir, fovs=[fov], channels=channels, dtype="int16"
+            )
+
+        # load segmentation labels in for fov
+        seg_labels = imread(os.path.join(seg_dir, fov + '_feature_0.tif'))
+
+        # # load segmentation labels in for fov
```
Comment (ngreenwald) [inline ark/phenotyping/som_utils.py:210]: leftover code

File: ark/utils/notebooks_test.py [hunk 7]
```diff
@@ -25,17 +25,17 @@ def _exec_notebook(nb_filename):
...
 
-def test_example_spatial_analysis():
-    _exec_notebook('example_spatial_analysis_script.ipynb')
+# def test_example_spatial_analysis():
+#     _exec_notebook('example_spatial_analysis_script.ipynb')
 
 
-def test_example_neighborhood_analysis():
-    _exec_notebook('example_neighborhood_analysis_script.ipynb')
+# def test_example_neighborhood_analysis():
```
Comment (ngreenwald) [inline ark/utils/notebooks_test.py:37]: ?

**Review by ngreenwald — CHANGES_REQUESTED** [review]

File: ark/utils/notebooks_test.py (hunk 7 above)
Comment (alex-l-kong) [reply ark/utils/notebooks_test.py:37]: Ah my bad, I commented them out to make it easier to run the FlowSOM notebook tests.

File: ark/phenotyping/som_utils.py (hunk 5 above)
Comment (alex-l-kong) [reply ark/phenotyping/som_utils.py:129]: Good idea.

**Review by ngreenwald — APPROVED** [review]
Looks great
